src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -31785,6 +31782,70 @@ static void M(C c, string? p1, string? p2)
             CompileAndVerify(comp2);
         }
 
+        [Fact]
+        public void NotNullIfNotNull_ByValueParameter_02()
+        {
+            var source = """
+                using System.Diagnostics.CodeAnalysis;
+
+                class C
+                {
+                    public static void M(string? p, [NotNullIfNotNull("p")] string? p2) => throw null!;
+                }
+                class D
+                {
+                    static void M(string p, string? p2)
+                    {
+                        C.M(p, p2);
+                        p.ToString();
+                        p2.ToString();
+                    }
+                }
+                """;
+
+            var comp = CreateNullableCompilation(new[] { source, NotNullIfNotNullAttributeDefinition });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NotNullIfNotNull_ByValueParameter_03()
+        {
+            // NotNullIfNotNull postcondition is applied once per call
+            // This means that successive calls may each change nullable state
+            var source = """
+                using System.Diagnostics.CodeAnalysis;
+
+                class C
+                {
+                    public static void M(string? p, [NotNullIfNotNull("p")] string? p2, [NotNullIfNotNull("p2")] string? p3) => throw null!;
+                }
+                class D
+                {
+                    static void M1(string p, string? p2, string? p3)
+                    {
+                        C.M(p, p2, p3);
+                        p.ToString();
+                        p2.ToString();
+                        p3.ToString(); // 1
+                    }
+                    static void M2(string p, string? p2, string? p3)
+                    {
+                        C.M(p, p2, p3);
+                        C.M(p, p2, p3);
+                        p.ToString();
+                        p2.ToString();
+                        p3.ToString();
+                    }
+                }
+                """;
+
+            var comp = CreateNullableCompilation(new[] { source, NotNullIfNotNullAttributeDefinition });
+            comp.VerifyDiagnostics(
+                // 0.cs(14,9): warning CS8602: Dereference of a possibly null reference.
```

Comment:
> // 0.cs(14,9): warning CS8602: Dereference of a possibly null reference. [](http://example.com/codeflow?start=15&length=73)

I find this warning unexpected. Since we figured out that `p2` is not null at the call site, then `p3` must be not null as well. Or are we somehow saying that we do not know if `p2` is not null at the call site, but the invocation ensures that it is not null after the call, by possibly adjusting the value? I didn't see a test for a scenario like that though.

Reply:
> Or are we somehow saying that we do not know if p2 is not null at the call site, but the invocation ensures that it is not null after the call, by possibly adjusting the value? I didn't see a test for a scenario like that though.

My goal for this PR is to have the attribute cause similar effects for both by-ref and by-value parameters. Adding test `NotNullIfNotNull_RefParameter_02`, to demonstrate that "closure" is not performed on by-ref parameters with NotNullIfNotNullAttribute. I think this is a case where we could do a more sophisticated and precise analysis in the future, but that giving a more "pessimistic" nullable state for now is not wrong.